lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -368,6 +376,119 @@ func (c *ContainerProxy) createNewAttachClientWithTimeouts(connectTimeout, respo
+	if err != nil {
+		if _, isa := err.(*interaction.ContainerResizeNotFound); isa {
+			cache.ContainerCache().DeleteContainer(vc.ContainerID)
+			return NotFoundError(vc.ContainerID)
+		}
+
+		// If we get here, most likely something went wrong with the port layer API server
+		return InternalServerError(err.Error())
+	}
+
+	return nil
+}
+
+// attacheStreams takes the the hijacked connections from the calling client and attaches
+// them to the 3 streams from the portlayer's rest server.
+// clStdin, clStdout, clStderr are the hijacked connection
+func (c *ContainerProxy) AttachStreams(ctx context.Context, vc *viccontainer.VicContainer, clStdin io.ReadCloser, clStdout, clStderr io.Writer, ca *backend.ContainerAttachConfig) error {
+	defer clStdin.Close()
+
+	// Cancel will close the child connections.
+	ctx, cancel := context.WithCancel(ctx)
+
+	var wg sync.WaitGroup
+	errors := make(chan error, 3)
+
+	// For stdin, we only have a timeout for connection.  There can be a long duration before
+	// the first entry so there is no timeout for response.
+	plClient, transport := createNewAttachClientWithTimeouts(attachConnectTimeout, 0, attachAttemptTimeout)
+	defer transport.Close()
+
+	if ca.UseStdin {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := copyStdIn(ctx, plClient, vc, clStdin, ca.DetachKeys)
+			if err != nil {
+				log.Errorf("container attach: stdin (%s): %s", vc.ContainerID, err.Error())
+			} else {
+				log.Infof("container attach: stdin (%s) done: %s", vc.ContainerID)
+			}
+
+			cancel()
+			errors <- err
+		}()
+	}
+
+	if ca.UseStdout {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := copyStdOut(ctx, plClient, attachAttemptTimeout, vc, clStdout)
+			if err != nil {
+				log.Errorf("container attach: stdout (%s): %s", vc.ContainerID, err.Error())
+			} else {
+				log.Infof("container attach: stdout (%s) done: %s", vc.ContainerID)
+			}
+
+			cancel()
+			errors <- err
+		}()
+	}
+
+	if ca.UseStderr {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := copyStdErr(ctx, plClient, vc, clStderr)
+			if err != nil {
+				log.Errorf("container attach: stderr (%s): %s", vc.ContainerID, err.Error())
+			} else {
+				log.Infof("container attach: stderr (%s) done: %s", vc.ContainerID)
+			}
+
+			cancel()
+			errors <- err
+		}()
+	}
+
+	// Wait for all stream copy to exit
+	wg.Wait()
+
+	log.Infof("container attach:  cleaned up connections to %s.", vc.ContainerID)
+	for err := range errors {
+		if err != nil {
+			// If we get here, most likely something went wrong with the port layer API server
+			// These errors originate within the go-swagger client itself.
+			// Go-swagger returns untyped errors to us if the error is not one that we define
+			// in the swagger spec.  Even EOF.  Therefore, we must scan the error string (if there
+			// is an error string in the untyped error) for the term EOF.
+
+			log.Errorf("container attach error: %s", err.Error())
```

Comment:
It looks like any error that happened is already logged. I wonder if there any reason to log first error that occurred and return it?